### PR TITLE
docs(code-conventions): update TypeScript conventions

### LINKED ISSUE #86: Update docs: new TS conventions
Update the [current convention doc](https://github.com/ethereum/ethereum-org-fork/blob/dev/docs/next-docs/code-conventions.md) with the new accepted TS conventions discussed in https://github.com/ethereum/ethereum-org-website/issues/11421

This should cover points 1, 2, and 4 from the proposal.

### PR DESCRIPTION
**Pull request by TylerAPfledderer**
## Description

Update the `code-conventions.md` doc with more defined conventions for TypeScript in this repo.

## Related Issue
Closes #86

### CODE REVIEW

**Review by pettinarip — APPROVED** [review]
LGTM! gj @TylerAPfledderer

**Comment by TylerAPfledderer** [comment]
@pettinarip I made an inconsistency here.

When mentioning intellisense for the direct props annotation, I'm showing the props destructured, but I don't show that anywhere else.

Currently afk at the time of this comment, but I would like to make it clear in this doc that props should be consistently destructured in this way.


NOT...

```tsx
const Component = (props: ComponentProps) => {
  const { title, ...rest } = props

  // Rest of the code
}
```

to maintain consistent intellisence DX.

Maybe this doesn't matter? I'm curious what you think about it, if you had any negative impact in trying to work with props by not having this kind of intellisense showing?

**Review by wackerow — APPROVED** [review]
Nice job @TylerAPfledderer! I agree with your comment about preferring to destructure props within the properties declaration and not by setting equal to props once inside the function. Feel free to update to make more explicit

**Review by wackerow — COMMENTED** [review]
If we're going to comment on how to destructure, may be worth noting the use of `...rest`

On that note, do we have a standard way we'd like to name the "rest props"? I've seen a variety of uses depending on the case (ie `...props`, `...rest`, `...restProps`, `...boxProps` etc). Do we want to align on this or are folks not picky?

**Comment by TylerAPfledderer** [comment]
> If we're going to comment on how to destructure, may be worth noting the use of `...rest`
> 
> On that note, do we have a standard way we'd like to name the "rest props"? I've seen a variety of uses depending on the case (ie `...props`, `...rest`, `...restProps`, `...boxProps` etc). Do we want to align on this or are folks not picky?

I'd think that if the destructuring is to occur at the props declaration, then I would prefer `...props` since the name is not going to be used otherwise.

**Comment by pettinarip** [comment]
@TylerAPfledderer 

Not a strong opinion from my side, I agree that would be better to be consistent if possible. `...props` or `...rest` are good to me.

**Comment by TylerAPfledderer** [comment]
@pettinarip @wackerow this is now updated with notation regarding the destructing object of props and the spread operator naming.

File: docs/next-docs/code-conventions.md [hunk 1]
```diff
@@ -23,3 +23,48 @@ TypeScript types should be defined inside `src/lib/types.ts` file. Note that som
...
+const Component = forwardRef<ComponentProps, "div">(
+  ({ title, label, ...props }, ref) => {
+    // Component code
+  }
+)
+```
+
+#### Prop Type Naming Convention
+
+For the props type signature use the naming convention `<ComponentName>Props` to provide an explicit name for the type. This is helpful for importing the signature to other files like page component and not force the user to alias the import every time, while providing readability.
```
Comment (wackerow) [inline docs/next-docs/code-conventions.md:58]: Do we want to always export the Props type?

Comment (wackerow) [reply docs/next-docs/code-conventions.md:58]: Or are we cool doing this on an individual basis if it is needed elsewhere?

**Review by wackerow — APPROVED** [review]

File: docs/next-docs/code-conventions.md [hunk 2]
```diff
@@ -23,3 +23,48 @@ TypeScript types should be defined inside `src/lib/types.ts` file. Note that som
...
+ *
+ * For the second argument of the generic types, you are declaring the primary element type that the component will render.
+ * This could be a `div`, `span`, `button`, etc. or a custom component (typeof Button) if said component is being used in the return.
+ */
+const Component = forwardRef<ComponentProps, "div">(
+  ({ title, label, ...props }, ref) => {
+    // Component code
+  }
+)
+```
```
Comment (TylerAPfledderer) [inline docs/next-docs/code-conventions.md:54]: @wackerow @pettinarip 
I added clarifying text for the user to know exactly what is applied to the generic type arguments for `forwardRef`. what do you think about this addition?

File: docs/next-docs/code-conventions.md (hunk 1 above)
Comment (TylerAPfledderer) [reply docs/next-docs/code-conventions.md:58]: I do not have a preference to argue for here.

File: docs/next-docs/code-conventions.md (hunk 2 above)
Comment (pettinarip) [reply docs/next-docs/code-conventions.md:54]: This is good for me. It is another use case where is useful to see our conventions applied to it.

File: docs/next-docs/code-conventions.md (hunk 1 above)
Comment (pettinarip) [reply docs/next-docs/code-conventions.md:58]: > Or are we cool doing this on an individual basis if it is needed elsewhere?

I think we were already doing that in the main repo, we were not enforcing the export.

File: docs/next-docs/code-conventions.md [hunk 3]
```diff
@@ -23,3 +23,48 @@ TypeScript types should be defined inside `src/lib/types.ts` file. Note that som
...
+
+For the props type signature use the naming convention `<ComponentName>Props` to provide an explicit name for the type. This is helpful for importing the signature to other files like page component and not force the user to alias the import every time, while providing readability.
+
+#### Directly annotate the props object
+
+**Do not use `React.FC`** and instead annotate the props object directly. `React.FC` implies the `children` prop, but this is not always desired when there is a component that should not accept this prop. `React.FC` also does not allow for use of Generic types, or use of Generic type when doing type guarding like function overloading. It is also not generally recommended to use and [was removed from the create-react-app template](https://github.com/facebook/create-react-app/pull/8177).
+
+A positive side-effect to directly annoting the props object is for IDE intellisense where you can view the props when hovering over the component name to see it's signature.
+
+i.e. `const Component: ({ label, title, ...rest }: ComponentProps) => React.JSX.Element`
```
Comment (pettinarip) [inline docs/next-docs/code-conventions.md]:
```suggestion
i.e. `const Component: ({ label, title, ...props }: ComponentProps) => React.JSX.Element`
```

**Pull request merged**